### PR TITLE
Refactored model API

### PR DESCRIPTION
**Pull request by subutai**

### CODE REVIEW

**Comment by subutai** [comment]
@BoltzmannBrain 

The API has been cleaned up. The following classes have now been transferred over to the refactored API:
```
classification_model.py
classify_document_fingerprint.py
classify_keywords.py
classify_htm.py
classify_fingerprint.py
```
All the above classes run for `hello_classification_model.py, simple_labels.py`, and `simple_query.py`.  

There were a couple of issues: 

1) returning detailed sorted results is slower than just returning the classification results. This is because the results have to be combined across tokens. As such I made this an option.

2) the logic for combining individual token results into document-level results needs work. The code is a bit repetitive between the classes, so we could simplify. More importantly I don't think the logic is ideal from an algorithm standpoint, so we might want to revisit later.  Nevertheless one implementation is in there.

This PR depends on a NuPIC change. I won't merge until after that has gone through, but a code review would be appreciated.

**Comment by subutai** [comment]
Depends on NuPIC PR https://github.com/numenta/nupic/pull/2931

File: htmresearch/frameworks/nlp/classification_model.py [hunk 1]
```diff
@@ -19,335 +19,183 @@
...
                numLabels=3,
-               modelDir="ClassificationModel"):
+               filterText=False,
+               **kwargs):
     """
-    If there are no labels set numLabels=0.
+    @param verbosity  (int) Verbosity level. Larger values lead to more
+                            printouts. 0 implies nothing will be printed.
+
+    @param numLabels  (int) The maximum number of categories.
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classification_model.py]: Previously `numLabels` was used as the max number of categories a model can infer. The use in this refactored model, however, implies `numLabels` is the total number of categories in the dataset. We should be explicit about the definition here.

File: htmresearch/frameworks/nlp/classification_model.py [hunk 2]
```diff
@@ -19,335 +19,183 @@
...
-  def prepText(text, preprocess=False):
-    """
-    Returns a list of the text tokens.
-
-    @param preprocess   (bool)    Whether or not to preprocess the text data.
+    @return  (numpy array) An array of size numLabels. Position i contains
+                           the likelihood that this token belongs to the
+                           i'th category. An array containing all zeros
+                           implies no decision could be made.
+             (list)        A list of sampleIds
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classification_model.py]: Do these correlate with the returned array of prototype distances?

File: htmresearch/frameworks/nlp/classification_model.py [hunk 3]
```diff
@@ -19,335 +19,183 @@
...
-  def tokenize(self, text, preprocess=False):
+  def inferDocument(self, document, returnDetailedResults=False,
+                    sortResults=True):
     """
-    Given a bunch of text (could be several sentences) return a single list
-    containing individual tokens.
+    Run inference on the model with this document and return classification
+    results, sampleIds and distances.  By default this routine will tokenize the
+    document and classify using these tokens. A reset is issued after inference.
+    Repeated sampleIds ARE removed from the results.
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classification_model.py]: Please note the classification method here is to get the single most likely category for each token and add 1 to the `categoryVotes` array.

File: htmresearch/frameworks/nlp/classification_model.py [hunk 4]
```diff
@@ -19,335 +19,183 @@
...
+    @param document (str)     The document to classify
+    @param returnDetailedResults
+                    (bool)    If True will return sampleIds and distances.
+                              This could slow things down depending on the
+                              number of stored patterns.
+    @param sortResults (bool) If true the list of sampleIds and distances
+                              will be sorted in order of increasing distances.
 
-    TODO: get rid of @staticmethod above and then we can remove this method.
+    @return  (numpy array) An array of size numLabels. Position i contains
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classification_model.py]: this needs updating

File: htmresearch/frameworks/nlp/classification_model.py [hunk 5]
```diff
@@ -19,335 +19,183 @@
...
+                           the likelihood that this token belongs to the i'th
+                           category. An array containing all zeros implies no
+                           decision could be made.
+             (list)        A list of unique sampleIds
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classification_model.py]: ... only returned if returnDetailedResults is True

File: htmresearch/frameworks/nlp/classification_model.py [hunk 6]
```diff
@@ -19,335 +19,183 @@
...
+             (numpy array) An array of distances from each stored sample
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classification_model.py]: ^^

File: htmresearch/frameworks/nlp/classification_model.py [hunk 7]
```diff
@@ -19,335 +19,183 @@
...
-    For models which define categories with bitmaps, log the categories (and
-    their relative distances) to a JSON specified by the dirName. The JSON will
-    contain the dict of category bitmaps, and if specified, dicts of label
-    references and category bitmap comparisons.
-    """
-    if not hasattr(self, "categoryBitmaps"):
-      raise TypeError("This model does not have category encodings compatible "
-                      "for logging.")
+    lastTokenIndex = len(tokenList) - 1
+    categoryVotes = numpy.zeros(self.numLabels)
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classification_model.py]: `numLabels` is used here to init an array with one entry for each possible category label, correct? See my comment in the constructor docstring about this.

File: htmresearch/frameworks/nlp/classification_model.py [hunk 8]
```diff
@@ -19,335 +19,183 @@
...
-    """
-    if self.numLabels == 0:
-      # No labels for classification, so populate labels with stand-ins
-      self.patterns = [{"recordNumber": i,
-                        "pattern": self.encodeSample(s[0]),
-                        "labels": numpy.array([-1]),
-                        "ID": s[2]}
-                       for i, s in samples.iteritems()]
+    if self.filterText:
+      sample = TextPreprocess().tokenize(text,
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classification_model.py]: better naming would be `preprocessedText`.

File: htmresearch/frameworks/nlp/classification_model.py [hunk 9]
```diff
@@ -502,3 +335,74 @@ def _deSerializeExtraData(self, extraDataDir):
...
+    # For each token run inference on the token and accumulate sum of distances
+    # from this token to all other sampleIds.
+    lastTokenIndex = len(tokenList) - 1
+    categoryVotes = numpy.zeros(self.numLabels)
+    distancesForEachId = {}
+    classifier = self.getClassifier()
+
+    for i, token in enumerate(tokenList):
+      votes, idList, distances = self.inferToken(token,
+                                     reset=int(i == lastTokenIndex),
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classification_model.py]: indent is off

File: htmresearch/frameworks/nlp/classification_model.py [hunk 10]
```diff
@@ -502,3 +335,74 @@ def _deSerializeExtraData(self, extraDataDir):
...
+                                     returnDetailedResults=True,
+                                     sortResults=False)
+
+      if votes.sum() > 0:
+        categoryVotes[votes.argmax()] += 1
+
+        # For each unique id, keep the minimum distance to this token
+        for j, sampleId in enumerate(idList):
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classification_model.py]: Is there ever a scenario where `votes` is all zeros but there are still IDs and distance values? What about unsupervised experiments (and Imbu) where `numLabels` = 0? Then these min distance calculations should not be w/in the current if block.

File: htmresearch/frameworks/nlp/classification_model.py [hunk 11]
```diff
@@ -502,3 +335,74 @@ def _deSerializeExtraData(self, extraDataDir):
...
+          )
+
+
+    # Put distance from each sampleId to this document into a numpy array
+    # ordered consistently with a list of sampleIds
+    sampleIdList = distancesForEachId.keys()
+    averageDistances = numpy.zeros(len(sampleIdList))
+    for i,sampleId in enumerate(sampleIdList):
+      averageDistances[i] = distancesForEachId.get(sampleId, numpy.inf)
+      # averageDistances[i] = (float(dist))/len(tokenList)
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classification_model.py]: rename this if not using the averages

File: htmresearch/frameworks/nlp/classify_document_fingerprint.py [hunk 12]
```diff
@@ -204,35 +199,46 @@ def trainText(self, token, labels, sequenceId=None, reset=0):
...
+    @param returnDetailedResults
+                    (bool)    If True will return sampleIds and distances
+                              This could slow things down depending on the
+                              number of stored patterns.
+    @param sortResults (bool) If true the list of sampleIds and distances
+                              will be sorted in order of increasing distances.
 
     @return  (numpy array) An array of size numLabels. Position i contains
-                           the likelihood that this sample belongs to the
+                           the likelihood that this token belongs to the
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_document_fingerprint.py]: an array of all zeros implies we have not yet run inference b/c the reset for end-of-document has not been flagged

File: htmresearch/frameworks/nlp/classify_document_fingerprint.py [hunk 13]
```diff
@@ -251,18 +258,43 @@ def classifyText(self, token, reset=0):
...
-        self.reset()
-
       if self.verbosity >= 2:
         print "Classifying document:",document
         self.printRegionOutputs()
 
       self.currentDocument = None
-      return self.classifierRegion.getOutputData("categoriesOut")[0:self.numLabels]
+      categoryVotes = self.classifierRegion.getOutputData(
+          "categoriesOut")[0:self.numLabels]
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_document_fingerprint.py]: Here you use `numLabels` to cap the number of inferred categories, but the base class `inferDocument()` uses it specify the total number of possible categories in the dataset. It's not clear why this is needed here.
And could just be `...[:self.numLabels]`

File: htmresearch/frameworks/nlp/classify_document_fingerprint.py [hunk 14]
```diff
@@ -251,18 +258,43 @@ def classifyText(self, token, reset=0):
...
+          sortedIndices = dist.argsort()
+          for i in sortedIndices:
+            idList.append(classifier.getPartitionId(i))
+          sortedDistances = dist[sortedIndices]
+          return categoryVotes, idList, sortedDistances
+
+        else:
+          idList = []
+          for i in range(len(dist)):
+            idList.append(classifier.getPartitionId(i))
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_document_fingerprint.py]: Replace this loop w/ `idList = [classifier.getPartitionId(i) for i in xrange(len(dist))]`.

File: htmresearch/frameworks/nlp/classify_fingerprint.py [hunk 15]
```diff
@@ -169,33 +94,20 @@ def trainText(self, token, labels, sequenceId=None, reset=0):
...
-        self.sampleReference.append(sequenceId)
-
-        # TODO: replace the need for sampleReference w/ partitionId.
-        # There is a bug in how partitionId is handled during infer if it is
-        # not passed in, so we won't pass it in for now (line 863 of
-        # KNNClassifier.py)
-        # self.classifier.learn(bitmap, label, isSparse=self.n,
-        #                       partitionId=sequenceId)
+        self.classifier.learn(bitmap, label, isSparse=self.encoder.n,
+                              partitionId=sampleId)
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_fingerprint.py]:
```
self.classifier.learn(
  bitmap, label, isSparse=self.encoder.n, partitionId=sampleId)
```

File: htmresearch/frameworks/nlp/classify_fingerprint.py [hunk 16]
```diff
@@ -204,26 +116,70 @@ def classifyText(self, token, reset=0, seed=42):
...
+    # Figure out format of returned results
+
+    if not returnDetailedResults:
+      # Return non-detailed results.
+      return inferenceResult, None, None
+
+    if not sortResults:
+      idList = []
+      for i in range(len(dist)):
+        idList.append(self.classifier.getPartitionId(i))
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_fingerprint.py]: `idList = [classifier.getPartitionId(i) for i in xrange(len(dist))]`

File: htmresearch/frameworks/nlp/classify_fingerprint.py [hunk 17]
```diff
@@ -204,26 +116,70 @@ def classifyText(self, token, reset=0, seed=42):
...
+      return inferenceResult, idList, dist
+
+    # Return sorted results
+    idList = []
+    sortedIndices = dist.argsort()
+    for i in sortedIndices:
+      idList.append(self.classifier.getPartitionId(i))
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_fingerprint.py]: (see list comprehension above)

File: htmresearch/frameworks/nlp/classify_fingerprint.py [hunk 18]
```diff
@@ -204,26 +116,70 @@ def classifyText(self, token, reset=0, seed=42):
...
+
+
+
+
+
+
+
+
+
+
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_fingerprint.py]: remove excess endlines

File: htmresearch/frameworks/nlp/classify_htm.py [hunk 19]
```diff
@@ -390,24 +143,24 @@ def reset(self):
...
+                             with this token.
+    @param sampleId   (int)  An integer ID associated with this token and its
                              sequence (document).
     @param reset      (int)  Should be 0 or 1. If 1, assumes we are at the
-                             beginning of a new sequence.
+                             end of the document.
     """
     for region in self.learningRegions:
       region.setParameter("learningMode", True)
+      region.setParameter("inferenceMode", True)
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_htm.py]: why?

File: htmresearch/frameworks/nlp/classify_htm.py [hunk 20]
```diff
@@ -418,27 +171,39 @@ def trainText(self, token, labels, sequenceId=None, reset=0):
...
                            implies no decision could be made.
+             (list)        A list of sampleIds
+             (numpy array) An array of distances from each stored sample
     """
     for region in self.learningRegions:
       region.setParameter("learningMode", False)
       region.setParameter("inferenceMode", True)
     sensor = self.sensorRegion.getSelf()
-    sensor.addDataToQueue(token, [None], -1, 0)
+    sensor.addDataToQueue(token, [None], sequenceId=-1, reset=0)
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_htm.py]: using the keyword would be helpful: `sensor.addDataToQueue(token, labels=[None], sequenceId=-1, reset=0)`

File: htmresearch/frameworks/nlp/classify_htm.py [hunk 21]
```diff
@@ -418,27 +171,39 @@ def trainText(self, token, labels, sequenceId=None, reset=0):
...
     self.network.run(1)
 
+    dist = self.classifierRegion.getSelf().getLatestDistances()
+
+    categoryLikelihoods = self.classifierRegion.getOutputData(
+        "categoriesOut")[0:self.numLabels]
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_htm.py]: see comment in htmresearch/frameworks/nlp/classify_document_fingerprint.py about this use of `numLabels`

File: htmresearch/frameworks/nlp/classify_htm.py [hunk 22]
```diff
@@ -446,7 +211,24 @@ def classifyText(self, token, reset=0):
...
+    partitionIdList = classifier.getPartitionIdPerPattern()
+    if not sortResults:
+      return categoryLikelihoods, partitionIdList, dist
+
+    # Sort results if requested
+    sortedIndices = dist.argsort()
+    sortedDistances = dist[sortedIndices]
+    sortedSampleIdList = []
+    for i in sortedIndices:
+      sortedSampleIdList.append(partitionIdList[i])
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_htm.py]: list comprehension would be better

File: htmresearch/frameworks/nlp/classify_keywords.py [hunk 23]
```diff
@@ -19,45 +19,30 @@
...
-
 class ClassificationModelKeywords(ClassificationModel):
   """
-  Class to run the survey response classification task with random SDRs.
-
-  From the experiment runner, the methods expect to be fed one sample at a time.
+  Class to run NLP classification task with random SDRs.
   """
-  # TODO: use nupic.bindings.math import Random
+  # TODO: use nupic.bindings.math import Random?
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_keywords.py]: I'd rather not, but this was requested by Scott a while back (I think for serialization concerns).

File: htmresearch/frameworks/nlp/classify_keywords.py [hunk 24]
```diff
@@ -67,176 +52,101 @@ def __init__(self,
...
+        idList = []
+        sortedIndices = dist.argsort()
+        for i in sortedIndices:
+          idList.append(self.classifier.getPartitionId(i))
+        sortedDistances = dist[sortedIndices]
+        return inferenceResult, idList, sortedDistances
+
+      else:
+        idList = []
+        for i in range(len(dist)):
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_keywords.py]: xrange in a list comprehension is better

File: htmresearch/frameworks/nlp/classify_keywords.py [hunk 25]
```diff
@@ -67,176 +52,101 @@ def __init__(self,
...
+          idList.append(self.classifier.getPartitionId(i))
+        return inferenceResult, idList, dist
+
+    else:
+        # Return non-detailed results.
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_keywords.py]: indent is off

File: htmresearch/frameworks/nlp/classify_keywords.py [hunk 26]
```diff
@@ -67,176 +52,101 @@ def __init__(self,
...
+        return inferenceResult, None, None
+
+  def _encodeToken(self, token):
+    """
+    Randomly encode an SDR of the input token. We seed the random number
+    generator such that a given string will yield the same SDR each time this
```
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_keywords.py]: "... will *return* the same..."
nitpicky, but this is not a generator :)

**Comment by BoltzmannBrain** [comment]
@subutai I gave it a read through, but have not yet tested the changes, including going through the example test scripts. I'll do this this weekend.

And have you merged in master? I notice that moving some files to the obsolete dir is included here, but I merged in these changes a few days ago in #398.

**Comment by subutai** [comment]
> And have you merged in master? I notice that moving some files to the obsolete dir is included here, but I merged in these changes a few days ago in #398.

Thanks for the review. Yes, I had merged in master. This PR includes some additional moves not included in #398 (based on our conversation a couple of days ago).

File: htmresearch/frameworks/nlp/classification_model.py (hunk 2 above)
Comment (subutai) [inline htmresearch/frameworks/nlp/classification_model.py]: > Do these correlate with the returned array of prototype distances?

Not always. Depends on whether the client requests sorted results or not.

File: htmresearch/frameworks/nlp/classification_model.py (hunk 4 above)
Comment (subutai) [inline htmresearch/frameworks/nlp/classification_model.py]: > this needs updating

In what way? I think the comment is still accurate for this API call.

File: htmresearch/frameworks/nlp/classification_model.py (hunk 10 above)
Comment (subutai) [inline htmresearch/frameworks/nlp/classification_model.py]: > Is there ever a scenario where votes is all zeros but there are still IDs and distance values?

Hmm, I don't think this is possible currently.

> What about unsupervised experiments (and Imbu) where numLabels = 0? Then these min distance calculations should not be w/in the current if block.

I think the current setup still requires at least one category. Not sure the KNNClassifier, etc. will work correctly without any categories.

File: htmresearch/frameworks/nlp/classify_htm.py (hunk 21 above)
Comment (subutai) [inline htmresearch/frameworks/nlp/classify_htm.py]: > see comment in htmresearch/frameworks/nlp/classify_document_fingerprint.py about this use of numLabels

I think this line is correct.

**Comment by subutai** [comment]
@BoltzmannBrain I finished pushing changes based on your review. I agreed with most of them but not all (we can discuss if you want). Thanks in particular for catching areas where list comprehension could be used.

**Pull request merged**